projection/src/main/scala/akka/projection/r2dbc/internal/R2dbcProjectionImpl.scala

```diff
@@ -299,16 +401,82 @@ private[projection] class R2dbcProjectionImpl[Offset, Envelope](
 
     implicit val executionContext: ExecutionContext = system.executionContext
     override val logger: LoggingAdapter = Logging(system.classicSystem, this.getClass)
+    private val log = LoggerFactory.getLogger(this.getClass)
 
     override def readPaused(): Future[Boolean] =
       offsetStore.readManagementState().map(_.exists(_.paused))
 
     override def readOffsets(): Future[Option[Offset]] =
       offsetStore.readOffset()
 
+    // Called from InternalProjectionState.saveOffsetAndReport
     override def saveOffset(projectionId: ProjectionId, offset: Offset): Future[Done] =
       offsetStore.saveOffset(offset)
 
+    override protected def saveOffsetAndReport(
+        projectionId: ProjectionId,
+        projectionContext: ProjectionContextImpl[Offset, Envelope],
+        batchSize: Int): Future[Done] = {
+      import R2dbcProjectionImpl.envToString
+      import R2dbcProjectionImpl.FutureDone
+      val envelope = projectionContext.envelope
+
+      if (offsetStore.isEnvelopeDuplicate(envelope)) {
+        // FIXME change to trace
+        log.debug("saveOffset filtering out duplicate: {}", envToString(envelope))
+        FutureDone
+      } else if (!offsetStore.wasSequenceNumberAccepted(envelope)) {
+        log.debug(
+          "saveOffset filtering out rejected sequence number (might be accepted later): {}",
+          envToString(envelope))
+        FutureDone
+      } else {
+        super.saveOffsetAndReport(projectionId, projectionContext, batchSize)
+      }
+    }
+
+    override protected def saveOffsetsAndReport(
+        projectionId: ProjectionId,
+        batch: Seq[ProjectionContextImpl[Offset, Envelope]]): Future[Done] = {
+      import R2dbcProjectionImpl.envToString
+      import R2dbcProjectionImpl.FutureDone
+
+      val acceptedContexts = batch.iterator.filterNot { ctx =>
+        val env = ctx.envelope
+        if (offsetStore.isEnvelopeDuplicate(env)) {
+          // FIXME change to trace
+          log.debug("saveOffsets filtering out duplicate: {}", envToString(env))
+          true
+        } else if (!offsetStore.wasSequenceNumberAccepted(env)) {
+          log.debug(
+            "saveOffsets filtering out rejected sequence number (might be accepted later): {}",
+            envToString(env))
+          true
+        } else {
+          false
+        }
+      }.toVector
+
+      if (acceptedContexts.isEmpty) {
+        FutureDone
+      } else {
+        val offsets = acceptedContexts.map(_.offset)
+        offsetStore
+          .saveOffsets(offsets)
+          .map { done =>
+            val batchSize = acceptedContexts.map { _.groupSize }.sum
+            val last = acceptedContexts.last
+            try {
+              statusObserver.offsetProgress(projectionId, last.envelope)
+            } catch {
+              case NonFatal(_) => // ignore
+            }
+            getTelemetry().onOffsetStored(batchSize)
+            done
+          }
+      }
+    }
```

Comment:
I opened up InternalProjectionState in https://github.com/akka/akka-projection/pull/600 to be able to override these. Must check the accepted here before saving offset, and must store all offsets, not only the latest.